drivers/meter/ade9430/ade9430.c

```diff
@@ -0,0 +1,360 @@
+	ret = ade9430_read(dev, watt_reg, &temp);
+	if (ret)
+		return ret;
+
+	dev->watt_val = temp * ADE9430_W_RES_UW / MICROWATT_PER_WATT;
+
+	return 0;
+}
+
+/**
+ * @brief Set User Energy use model.
+ * @param dev - The device structure.
+ * @param model - The User Energy use model to be set.
+ * @param value - value specific to the user energy use model, it can be either
+ * 		1 for reading with reset, number of half line cycles or number
+ * 		of samples.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9430_set_egy_model(struct ade9430_dev *dev, enum ade9430_egy_model model,
+			  uint16_t value)
+{
+	int ret;
+	uint16_t egy_ld_accum, egy_tmr_mode, rd_rst_en, egy_pwr_en;
+
+	switch (model) {
+	case ADE9430_EGY_WITH_RESET:
+		/* For reading energy with reset the value should be 1 */
+		if (value != 1)
+			return -EINVAL;
+		egy_ld_accum = 0;
+		egy_tmr_mode = 0;
+		rd_rst_en = 1;
+		egy_pwr_en = 1;
+		break;
+	case ADE9430_EGY_HALF_LINE_CYCLES:
+		egy_ld_accum = 1;
+		egy_tmr_mode = 1;
+		rd_rst_en = 0;
+		egy_pwr_en = 1;
+		break;
+	case ADE9430_EGY_NR_SAMPLES:
+		egy_ld_accum = 1;
+		egy_tmr_mode = 0;
+		rd_rst_en = 0;
+		egy_pwr_en = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = ade9430_write(dev, ADE9430_REG_RUN, 0);
+	if (ret)
+		return ret;
+
+	ret = ade9430_write(dev, ADE9430_REG_EP_CFG,
+			    no_os_field_prep(ADE9430_EGY_LD_ACCUM, egy_ld_accum) |
+			    no_os_field_prep(ADE9430_EGY_TMR_MODE, egy_tmr_mode) |
+			    no_os_field_prep(ADE9430_RD_RST_EN, rd_rst_en) |
+			    no_os_field_prep(ADE9430_EGY_PWR_EN, egy_pwr_en));
+	if (ret)
+		return ret;
+
+	ret = ade9430_write(dev, ADE9430_REG_EGY_TIME, value);
+	if (ret)
+		return ret;
+
+	return ade9430_write(dev, ADE9430_REG_RUN, 1);
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9430_init(struct ade9430_dev **device,
+		 struct ade9430_init_param init_param)
+{
+	struct ade9430_dev *dev;
+	uint32_t chip_id;
+	int ret;
+
+	dev = (struct ade9430_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* SPI Initialization*/
+	ret = no_os_spi_init(&dev->spi_desc, init_param.spi_init);
+	if (ret)
+		goto error_dev;
+
+	ret = ade9430_update_bits(dev, ADE9430_REG_CONFIG1, ADE9430_SWRST,
+				  no_os_field_prep(ADE9430_SWRST, 1));
+	if (ret)
+		goto error_spi;
+
+	no_os_mdelay(100);
+
+	ret = ade9430_update_bits(dev, ADE9430_REG_CONFIG1, ADE9430_SWRST,
+				  no_os_field_prep(ADE9430_SWRST, 0));
+	if (ret)
+		goto error_spi;
+
+	/** Use a valid register with default value different from 0*/
+	ret = ade9430_read(dev, ADE9430_REG_CONFIG5, &chip_id);
+	if (ret)
+		goto error_spi;
+
+	if (chip_id != ADE9430_CHIP_ID)
+		goto error_spi;
+
+	/* Enable Temperature Sensor */
+	ret = ade9430_update_bits(dev, ADE9430_REG_TEMP_CFG, ADE9430_TEMP_EN,
```

Comment:
This is an optional change:

I can see why you always want to enable temperature (you have an API for reading temperature that assumes temperature sensor is enabled), but it'd probably be better to specify this through init_param.

Reply:
Sure, will do.